torch/csrc/PyInterpreter.cpp

```diff
@@ -562,7 +562,133 @@ static void set_tensor_attr_with_capsule(
       getPyInterpreter(), /*ignore_hermetic_tls=*/false);
   TORCH_CHECK(
       mb_obj.has_value(), "Tensor subclass's PyInterpreter has no value");
-  py::handle(mb_obj.value()).attr(attr_name) = capsule;
+  auto obj = mb_obj.value();
+  py::handle(obj).attr(attr_name) = capsule;
+}
+
+// Note [Tensor Subclass custom size/stride caching strategy]
+// Tensor subclasses can use __torch_dispatch__ to override size/stride calls.
+// However, this presents a problem:
+// (1) When you return a custom (maybe symbolic) size/stride
+//     from python, we need to stash this fresh vector of ints/symints
+//     somewhere so that it has the same lifetime as the tensor.
+// (2) If the subclass experiences a metadata mutation,
+//     this stashed vector is no longer valid, so we need to allocate a fresh
+//     buffer to store the new sizes the next time someone asks for them.
+//
+// We handle this in the same way that `TensorImpl::sizes_default()`
+// handles its buffer: we simply reallocate the buffer whenever
+// the number of dimensions changes due to a resize.
+// Notable, we do *not* reallocate the buffer if the values changed,
+// but the number of dimensions stayed the same (e.g. `.transpose_()`).
+template <typename T>
+static c10::ArrayRef<T> get_set_cached_attr(
+    const c10::TensorImpl* tensor,
+    const char* base_attr_name,
+    const py::object& obj) {
+  c10::optional<PyObject*> mb_obj =
+      tensor->pyobj_slot()->check_pyobj(getPyInterpreter());
+  TORCH_CHECK(
+      mb_obj.has_value(), "Tensor subclass's PyInterpreter has no value");
+  auto tensor_obj = mb_obj.value();
+  auto buffer_len_attr_name = std::string(base_attr_name) + std::string("_len");
+
+  bool is_buffer_allocated = false;
+  size_t curr_size = 0;
+  if (PyObject_HasAttrString(tensor_obj, buffer_len_attr_name.c_str())) {
+    auto buffer_pyobj =
```

Comment:
nit: it's weird to call this `buffer_pyobj` since it's not actually the buffer lol